nautobot/extras/signals.py

```diff
@@ -53,6 +54,57 @@
 #
 
 
+def _cache_obj_data_in_change_context(action, sender, instance):
+    """
+    If this is an existing object that should have a change log entry,
+    we need to retrieve the existing object from the database and cache its data in the change context.
+    """
+    # We are caching the before object data in the change context so that it can be used later
+    change_context = change_context_state.get()
+
+    # Do nothing if the change_contex is None
+    if change_context is None:
+        return
+
+    # Is this an object without a change log?
+    if not hasattr(instance, "to_objectchange"):
+        return
+
+    # Does this object type not persist in the database?
+    if not hasattr(instance, "present_in_database"):
+        return
+
+    # Is this a new object?
+    if not instance.present_in_database:
+        return
+
+    # Does this object already have changes?
+    if ObjectChange.objects.filter(changed_object_id=instance.pk).exists():
+        return
+
+    # Retrieve the existing object from the database.
+    if action == ObjectChangeActionChoices.ACTION_UPDATE:
+        instance = sender.objects.get(pk=instance.pk)
+
+    # The context manager retrieves all changes by request ID. If we use the
+    # same request ID for this change, multiple events will be published rather
+    # than a single change. This is wildly inelegant, but scopes the hack to
+    # just this one file.
+    change = instance.to_objectchange(
+        action=action,
+    )
+    change.request_id = uuid.uuid4()
+    change.user = change_context.get_user(instance)
+    # cache the previous object data in the change context
+    if change_context.pre_object_data is None:
+        change_context.pre_object_data = {}
+    if change_context.pre_object_data_v2 is None:
+        change_context.pre_object_data_v2 = {}
+    change_context.pre_object_data[str(instance.pk)] = change.object_data
+    change_context.pre_object_data_v2[str(instance.pk)] = change.object_data_v2
```

Comment:
Do we have any potential issue here with an object being updated multiple times in the same change context? Might want to use `.setdefault()` here and/or wrap this whole block in a `if str(instance.pk) not in change_context.pre_object_data...` to avoid overwriting an existing cache in such a case?